### DISCUSSION TITLE
Inserting to an empty array when updating the cache using cache.updateQuery is not reflected to the actual state of the query

### DISCUSSION THREAD

**Discussion by EssamKayal96** [urql-graphql/urql#2986] [Q&A]
### Hello all, first of all sorry but I am new to urql,

**I am trying to update the cache using cache.updateQuery for a field that is nested** 

**the data initially looks like that,**

![image](https://user-images.githubusercontent.com/63129972/221916725-651709ca-40cb-4252-b045-7a07fdbc84c5.png)

**and whenever the user wants to insert I will update it based on the user_id which is given in the _arg object** 

**what's happening here is that the data is altered inside the cacheExchange but not in the state**

```cacheExchange({
  updates: {
    Mutation: {
      insert_app_user_offline_session_one(result, _args, cache, _info) {
        const courseId = window.location.pathname.split('/')[5];
        const sessionId = window.location.pathname.split('/')[7];
        const prev = cache
          .inspectFields('Query')
          .filter((field) => field.fieldName === 'app_regular_student');
        if (prev.length > 0) {
          cache.updateQuery(
            {
              query: AbsenceRecord,
              variables: {
                offine_session_template_id: courseId,
                offline_session_id: sessionId,
                limit: prev.at(-1)?.arguments?.limit,
                offset: prev.at(-1)?.arguments?.offset,
                order_by: prev.at(-1)?.arguments?.order_by,
              },
            },
            (data) => {
              const index = data?.app_regular_student.findIndex(
                (item) => item?.id === _args?.object?.regular_student_id
              );
              if (index !== -1)
                data.app_regular_student[index].user_offline_sessions = [
                  ...data.app_regular_student[index].user_offline_sessions,
                  result?.insert_app_user_offline_session_one,
                ];
              console.log(data);
              return data;
            }
          );
        }
      },
    },
  },
})```

**so when I monitor the log I see that the data before returning it is altered and the user_offline_sessions is not empty anymore**

**but when I try to log the data inside the component is not changed**


**any help please, and thanks in advance**

**Comment by kitten** [comment]
> but when I try to log the data inside the component is not changed

Ok, so you've posted the code for your updater in Graphcache, but as you're saying, that's working, and you validated this already, as far as I can tell? So, when you say "but not in the state" what does that look like? Which UI bindings are you using and what does your state there look like?

---

it's a little hard to read your snippets btw. [GitHub has some docs and advice on that](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) but basically, there's no need to use titles or to bold all text in your comments, and you can annotate your code, e.g. as `js` ✌️

**Reply by EssamKayal96** [reply]
Thanks for your reply, I am using urql within React.js

The idea here after the updater is done, inside the callback function of the cache.updateQuery I print the data after making some changes to it and the data seems to be changed and I get what I want, but the data is not being changed inside the JSX Component.


and I have already worked with changing the state before but not adding to an empty array, changing some props of the state and pushing to a non empty array, but when it comes to insert to an empty array it is not changed.

**Reply by JoviDeCroock** [reply]
Could it be that you are observing a different query, i.e. different orderBy or whatnot? Because you are only applying it to the first one 😅
